### PR TITLE
Add an ability to exclude particular spans from being recorded in traces

### LINKED ISSUE #4345: Setup Otel exporter for FerretDB
### What should be done?

We have some code that setups Otel exporter for integration tests and in envtool. We should refactor it and use it for setting Otel exporter in the FerretDB process itself. It should be configurable with `--test-XXX` flags.

We should not "fix" traces in this issue.

### Where?

TBD

### Definition of Done

TBD

- handler updated;
- all backends updated;
- unit tests added/updated;
- integration/compatibility tests added/updated;
- spot refactorings done;
- user documentation updated or an issue to create documentation created;
- something else?

### PR DESCRIPTION
**Pull request by rumyantseva**
# Description

Closes #4345.

The implementation is based on this discussion: https://github.com/open-telemetry/opentelemetry-go/issues/4718.

Also, if it were possible to propagate parent's attributes to all the children's spans automatically, the algorithm of span exclusion would be much easier. But there isn't an option for it, so we need to find all the children while exporting spans.

Please leave your feedback if you see that something can be simplified.

## Readiness checklist

- [x] I added/updated unit tests (and they pass).
- [x] I added/updated integration/compatibility tests (and they pass).
- [x] I added/updated comments and checked rendering.
- [ ] I made spot refactorings.
- [ ] I updated user documentation.
- [ ] I ran `task all`, and it passed.
- [x] I ensured that PR title is good enough for the changelog.
- [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Milestone (`Next`), Labels, Project and project's Sprint fields.
- [x] I marked all done items in this checklist.

### CODE REVIEW

File: internal/util/observability/observability.go [hunk 1]
```diff
@@ -111,3 +121,61 @@ func (ot *OtelTracer) Run(ctx context.Context) {
...
+			}
+
+			delete(toProcess, parentID)
+		}
+	}
+
+	var filteredSpans []otelsdktrace.ReadOnlySpan
+	for _, sps := range toProcess {
+		filteredSpans = append(filteredSpans, sps...)
+	}
```
Comment (chilagrow) [inline internal/util/observability/observability.go:173]:
```suggestion
filteredSpans := maps.Values(toProcess) 
```

File: internal/util/observability/observability.go [hunk 2]
```diff
@@ -111,3 +121,61 @@ func (ot *OtelTracer) Run(ctx context.Context) {
...
+// ExporterWithFilter is a span exporter that filters out spans with ExclusionAttribute and all their subspans.
+type ExporterWithFilter struct {
+	exporter otelsdktrace.SpanExporter
+}
+
+// ExclusionAttribute is the attribute that marks spans to be excluded.
+var ExclusionAttribute = attribute.KeyValue{Key: "excluded", Value: attribute.BoolValue(true)}
+
+// ExportSpans implements otelsdktrace.SpanExporter interface.
+func (e *ExporterWithFilter) ExportSpans(ctx context.Context, spans []otelsdktrace.ReadOnlySpan) error {
```
Comment (chilagrow) [inline internal/util/observability/observability.go:134]: Let me see if I understand what we are trying to do. Do we want to stop trace for envtool? I'm trying to understand the implementation of `ExclusionAttribute` to our usage, is it only for envtool? Currently, this is pretty tricky to understand.

Could a simpler solution be used? In https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#example-SpanProcessor-Filtered
there is `InstrumentationBlacklist` example which drops span with specified InstrumentationScope name.

**Review by chilagrow — COMMENTED** [review]
I have a question about how this implementation is expected to be used. Why is it important to hide particular span? :thinking:

File: internal/util/observability/observability.go [hunk 3]
```diff
@@ -111,3 +121,61 @@ func (ot *OtelTracer) Run(ctx context.Context) {
 
 	ot.l.Info("OTLP tracer stopped.")
 }
+
+// ExporterWithFilter is a span exporter that filters out spans with ExclusionAttribute and all their subspans.
+type ExporterWithFilter struct {
```
Comment (AlekSi) [inline internal/util/observability/observability.go:126]: Please unexport identifiers that should not be exported

**Comment by rumyantseva** [comment]
I close this PR as #4423 provides a simpler solution.

**Pull request closed without merging**